File: frogger.hs
-- https://open.kattis.com/problems/1dfroggereasy

frogger :: [Int] -> Int -> Int -> Int -> Int -> [Int] -> (String, Int)
frogger nums n i m h visited
    | k == m            = ("magic", h)
    | i `elem` visited  = ("cycle", h)
    | i < 0             = ("left", h)
    | i >= n            = ("right", h)
    | otherwise         = frogger nums n (i + k) m (h + 1) (visited ++ [i])
    where
        k = nums !! i

main :: IO ()
main = do
    line <- getLine
    let [n, s, m] = map read (words line) :: [Int]
    line <- getLine
    let nums = map read (words line) :: [Int]
    let (fate, h) :: (String, Int) = frogger nums n (s - 1) m 0 []
    putStrLn fate
    print h

File: frogger.py
n, s, m = map(int, input().split())
nums = [int(s) for s in input().split()]

i = s - 1
h = 0
visited = [False] * n
while True:

    k = nums[i]

    # on magic square
    if k == m:
        print('magic')
        break

    # square already visited; cycle
    if visited[i]:
        print('cycle')
        break

    # mark square as visited 
    visited[i] = True

    # jump k steps right/left
    i += k
    h += 1

    # fell off left edge
    if i < 0:
        print('left')
        break

    # fell off right edge
    if i >= n:
        print('right')
        break


print(h)